File: src/main/java/com/github/projectsandstone/api/event/entity/EntityDespawnEvent.java
/*
 *      SandstoneAPI - Minecraft Server Modding API
 *
 *         The MIT License (MIT)
 *
 *      Copyright (c) Sandstone <https://github.com/ProjectSandstone/>
 *      Copyright (c) contributors
 *
 *
 *      Permission is hereby granted, free of charge, to any person obtaining a copy
 *      of this software and associated documentation files (the "Software"), to deal
 *      in the Software without restriction, including without limitation the rights
 *      to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *      copies of the Software, and to permit persons to whom the Software is
 *      furnished to do so, subject to the following conditions:
 *
 *      The above copyright notice and this permission notice shall be included in
 *      all copies or substantial portions of the Software.
 *
 *      THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *      IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *      FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *      AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *      LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *      OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *      THE SOFTWARE.
 */
package com.github.projectsandstone.api.event.entity;

import com.github.projectsandstone.api.entity.living.LivingEntity;
import com.github.projectsandstone.eventsys.ap.Factory;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static com.github.projectsandstone.api.event.Constants.FACTORY_CLASS;

/**
 * Fired when an {@link com.github.projectsandstone.api.entity.Entity} despawn.
 */
@Factory(FACTORY_CLASS)
public interface EntityDespawnEvent extends EntityEvent {

    @Factory(FACTORY_CLASS)
    interface Death extends EntityDespawnEvent, LivingEntityEvent {

        /**
         * Gets killer of entity.
         *
         * @return Killer of entity.
         */
        @Nullable
        LivingEntity getKiller();

        @NotNull
        @Override
        LivingEntity getEntity();
    }
}

File: src/main/kotlin/com/github/projectsandstone/api/event/entity/EntityDespawnEvent.kt
/*
 *      SandstoneAPI - Minecraft Server Modding API
 *
 *         The MIT License (MIT)
 *
 *      Copyright (c) Sandstone <https://github.com/ProjectSandstone/>
 *      Copyright (c) contributors
 *
 *
 *      Permission is hereby granted, free of charge, to any person obtaining a copy
 *      of this software and associated documentation files (the "Software"), to deal
 *      in the Software without restriction, including without limitation the rights
 *      to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *      copies of the Software, and to permit persons to whom the Software is
 *      furnished to do so, subject to the following conditions:
 *
 *      The above copyright notice and this permission notice shall be included in
 *      all copies or substantial portions of the Software.
 *
 *      THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *      IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *      FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *      AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *      LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *      OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *      THE SOFTWARE.
 */
package com.github.projectsandstone.api.event.entity

import com.github.projectsandstone.api.entity.living.LivingEntity
import com.github.koresframework.eventsys.ap.Factory

import com.github.projectsandstone.api.event.Constants.FACTORY_CLASS

/**
 * Fired when an [com.github.projectsandstone.api.entity.Entity] despawn.
 */
@Factory(FACTORY_CLASS)
interface EntityDespawnEvent : EntityEvent {

    @Factory(FACTORY_CLASS)
    interface Death : EntityDespawnEvent, LivingEntityEvent {

        /**
         * Killer of entity.
         */
        val killer: LivingEntity?

        override val entity: LivingEntity
    }
}
